### ISSUE TITLE
Register URL handler for code scheme

### ISSUE THREAD

**Issue by kalikiana** [coder/code-server#1571]
code-server should be able to handle the `code` scheme, so URLs like `vscode:extension/kayteh.plastic-slate` such as used to install themes via a **Open in VSCode** button e.g. on the [Plastic Slate theme](https://vscodethemes.com/e/kayteh.plastic-slate) page on [VSCodeThemes](https://vscodethemes.com).

**Comment by nhooyr** [comment]
Given code-server doesn't run as an app locally, I'm not sure how we'd register for and handle such URLs.

**Comment by nhooyr** [comment]
Neat, didn't know that existed, thanks for sharing.

**Comment by bilogic** [comment]
I would have thought this is handled by vscode, is it still on the roadmap?

**Comment by code-asher** [comment]
It is not on the roadmap, and I am not sure if VS Code has added support for this.  Does anyone know if it works in Codespaces?

**Comment by TamiTakamiya** [comment]
I am also interested in this feature. Though I have not tried it out by myself yet, according to [the Codespaces documentation](https://code.visualstudio.com/api/advanced-topics/remote-extensions#callbacks-and-uri-handlers)  and [this article](https://www.eliostruyf.com/authentication-provider-work-github-gitpod-codespaces/) on a similar topic, Codespaces has a special endpoint (`https://*.github.dev/extension-auth-callback`) for auth callback.  Probably the endpoint forwards callbacks to individual URL handlers, which are written for regular VSCode.

**Comment by TamiTakamiya** [comment]
Update on this topic: For last few weeks, I have been working on enabling OAuth2 authorization code flow from our VSCode extension running in code-server and found `vscode.env.asExternalUri` could be used for the redirect URI as described in [the Codespaces documentation](https://code.visualstudio.com/api/advanced-topics/remote-extensions#callbacks-and-uri-handlers), i.e., instead of using the raw code scheme like `vscode://mycompany.feature`, use an external URI obtained from `vscode.env.asExternalUri('vscode://mycompany.feature')`.  The change needed to go in our VS Code extension code and [this](https://github.com/ansible/vscode-ansible/pull/1012/files) shows what I have changed.

While the changes on VS extension side were relatively simple, we needed to make some changes on the server side as well for (1) supporting wildcard in supported redirect URIs, and (2) making sure URI-encoded query parameters are processed properly. Since our server code is stored in a private repo, I cannot show those changes here.  Anyway, I just wanted to say no changes were needed on code-server to make it work.

**Comment by bilogic** [comment]
@TamiTakamiya 

Is there a way to call `vscode.env.asExternalUri` in my code-server without an extension?

**Comment by TamiTakamiya** [comment]
> @TamiTakamiya
> 
> Is there a way to call `vscode.env.asExternalUri` in my code-server without an extension?

Hmm... I am not sure if it is possible... 

I wrote [a very primitive VS Code extension](https://github.com/TamiTakamiya/asExternalUri-sample/) based on [the "Hello World" sample](https://code.visualstudio.com/api/get-started/your-first-extension) on VS Code Documentation.  Does it satisfy your needs?

The repo contains [the vsix file](https://github.com/TamiTakamiya/asExternalUri-sample/blob/main/asExternalUri-sample-0.0.1.vsix) and [README](https://github.com/TamiTakamiya/asExternalUri-sample/blob/main/README.md) contains a video on how to install/use it.

**Comment by bilogic** [comment]
@TamiTakamiya 

1. My project has a page that has a link `vscode://file/path/to/my/file.txt`
2. So I follow your video and entered `vscode://file`
3. However, the file did not open in my code-server when I click the link

Any ideas?

**Comment by TamiTakamiya** [comment]
@bilogic So far I have opened external URIs with our VSCode extensions.

When code-server is used, I think the scheme needs to be set to `code-oss:` instead of `vscode:`.  By specifying `code-oss://file/home/ttakamiy/test.txt` on my sample extension, I got the following external URI:

http://localhost:8080/stable-d7a2b4936af1bfd80cb96f2567af68badc2325e3/callback?vscode-reqid%3D1%26vscode-scheme%3Dcode-oss%26vscode-authority%3Dfile%26vscode-path%3D%252Fhome%252Fttakamiy%252Ftest.txt

but this does not open the file in the editor... I am not sure if this is supposed to work.

**Comment by bilogic** [comment]
@TamiTakamiya thanks, so I have to specify the full path in your sample and not just the front part? Let me try again.

**Comment by bilogic** [comment]
Also related to https://github.com/coder/code-server/discussions/3891

**Comment by bilogic** [comment]
revisiting, this is definitely doable

1. https://developer.chrome.com/docs/web-platform/best-practices/url-protocol-handler
2. https://protocol-handler.glitch.me/ - install PWA and register handler
3. F12 to open your dev-tools, Applications, Manifest, Protcol hander
4. Enter as example `web+coffee://americano`, it will open the PWA installed in 1
